src/java/org/apache/cassandra/db/monitoring/SaiSlowLog.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.monitoring;
+
+import java.beans.ConstructorProperties;
+import java.util.Comparator;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Objects;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.index.sai.QueryContext;
+import org.apache.cassandra.tracing.Tracing;
+
+public class SaiSlowLog
+{
+    private static final Logger logger = LoggerFactory.getLogger(SaiSlowLog.class);
+
+    private final AtomicInteger numSlowestQueries = new AtomicInteger();
+    // a structure that keeps the N slowest queries - don't use getter/setter so that it's not exposed over JMX
+    private final SaiSlowestQueriesQueue slowestQueries;
+    // KATE: make this false by default and decide on yaml property, -D flag, or JMX to enable; hook setEnabled() to change
+    private final AtomicBoolean enabled = new AtomicBoolean(true);
```

Comment:
No need to allow to disable this at runtime.
Simpler is better

Reply:
So, it is a -D flag or yaml property. However, I would like to say that it is a matter of adding two more methods to make this particular flag runtime configurable. It does not also require extensive testing. My preference is always to have some extremely quick way to start/stop using new things.

Reply:
> quick way to start/stop using new things

I understand your point, but if you want to go this way then we have to add an HTTP endpoint in CNDB in order to toggle this flag.

Reply:
As we talked in Slack, yes, I am aware we will need HTTP point to be added when we are done here. I haven't proposed it yet as first I wanted to be sure what we really want. The PR is for early feedback